src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -2014,4 +2138,594 @@ void CorProfiler::GetAssemblyAndSymbolsBytes(BYTE** pAssemblyArray, int* assembl
+        continue;
+      }
+
+      // Compare each mdMethodDef argument type to the instrumentation target
+      bool argumentsMismatch = false;
+      const auto methodArguments = functionInfo->method_signature.GetMethodArguments();
+      Debug("Comparing signature for method: ", integration.replacement.target_method.type_name, ".", integration.replacement.target_method.method_name);
+      for (unsigned int i = 0; i < numOfArgs; i++) {
+        const auto argumentTypeName = methodArguments[i].GetTypeTokName(metadata_import);
+        const auto integrationArgumentTypeName = integration.replacement.target_method.signature_types[i + 1];
+        Debug("  -> ", argumentTypeName, " = ", integrationArgumentTypeName);
+        if (argumentTypeName != integrationArgumentTypeName) {
+          argumentsMismatch = true;
+          break;
+        }
+      }
+      if (argumentsMismatch) {
+        Debug("The caller for the methoddef: ", integration.replacement.target_method.method_name, " doesn't have the right type of arguments.");
+        delete functionInfo;
+        enumIterator = ++enumIterator;
+        continue;
+      }
+
+      // As we are in the right method, we gather all information we need and stored it in to the ReJIT handler.
+      auto moduleHandler = rejit_handler->GetOrAddModule(module_id);
+      moduleHandler->SetModuleMetadata(module_metadata);
+      auto methodHandler = moduleHandler->GetOrAddMethod(methodDef);
+      methodHandler->SetFunctionInfo(functionInfo);
+      methodHandler->SetMethodReplacement(new MethodReplacement(integration.replacement));
+
+      // Store module_id and methodDef to request the ReJIT after analyzing all integrations.
+      vtModules.push_back(module_id);
+      vtMethodDefs.push_back(methodDef);
+      
+      bool caller_assembly_is_domain_neutral = runtime_information_.is_desktop() && corlib_module_loaded && module_metadata->app_domain_id == corlib_app_domain_id;
+
+      Info("Enqueue for ReJIT [ModuleId=", module_id,
+           ", MethodDef=", HexStr(&methodDef, sizeof(mdMethodDef)), 
+           ", AppDomainId=", module_metadata->app_domain_id,
+           ", IsDomainNeutral=", caller_assembly_is_domain_neutral,
+           ", Assembly=", module_metadata->assemblyName, 
+           ", Type=", caller.type.name, 
+           ", Method=", caller.name, 
+           ", Signature=", caller.signature.str(),
+           "]");
+      
+      enumIterator = ++enumIterator;
+    }
+  }
+
+  // Request the ReJIT for all integrations found in the module.
+  if (vtMethodDefs.size() > 0) {
+    Info("Requesting ReJIT for ", vtMethodDefs.size(), " methods.");
+    this->info_->RequestReJIT((ULONG)vtMethodDefs.size(), vtModules.data(), vtMethodDefs.data());
+  }
+
+  // We return the number of ReJIT requests
+  return vtMethodDefs.size();
+}
+
+/// <summary>
+/// Rewrite the target method body with the calltarget implementation. (This is function is triggered by the ReJIT handler)
+/// </summary>
```

Comment:
This method is pretty important so I feel like no amount of documentation for this algorithm will be too much. Here's an idea of what I would put here:

```
/// <summary>
/// Rewrite the target method body with the calltarget implementation. (This is function is triggered by the ReJIT handler)
/// Resulting code structure:
/// 
/// - Add locals for TReturn (if non-void method), CallTargetState, CallTargetReturn/CallTargetReturn<TReturn>, Exception
/// - Initialize locals
///
/// try
/// {
///   try
///   {
///     try
///     {
///       - Invoke BeginMethod with object instance (or null if static method) and original method arguments
///       - Store result into CallTargetState local
///     }
///     catch
///     {
///       - Invoke LogException(Exception)
///     }
///
///     - Execute original method instructions
///       * All RET instructions are replaced with a LEAVE_S. If non-void method, the value on the stack is first stored in the TReturn local.
///   }
///   catch (Exception)
///   {
///     - Store exception into Exception local
///     - throw
///   }
/// }
/// finally
/// {
///   try
///   {
///     - Invoke EndMethod with object instance (or null if static method), TReturn local (if non-void method), CallTargetState local, and Exception local
///     - Store result into CallTargetReturn/CallTargetReturn<TReturn> local
///     - If non-void method, store CallTargetReturn<TReturn>.GetReturnValue() into TReturn local
///   }
///   catch
///   {
///     - Invoke LogException(Exception)
///   }
/// }
///
/// - If non-void method, load TReturn local
/// - RET
/// </summary>
```

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/1007/commits/9e7bcb3955022de47db4e80e29c33db9f64b69c2